lib/chainweb/cut.ex

```diff
@@ -0,0 +1,134 @@
+defmodule Kadena.Chainweb.Cut do
+  @moduledoc """
+  `Cut` struct definition.
+  """
+  @behaviour Kadena.Chainweb.Type
+
+  @type hashes :: map() | list()
+  @type height :: non_neg_integer()
+  @type id :: String.t()
+  @type instance :: String.t()
+  @type origin :: map() | nil
+  @type weight :: String.t()
+  @type chain_id :: 0..19 | String.t() | atom()
+  @type error :: {:error, Keyword.t()}
+  @type result :: t() | error()
+  @type validation :: {:ok, map() | chain_id()} | error()
+
+  @type t :: %__MODULE__{
+          hashes: hashes(),
+          height: height(),
+          weight: weight(),
+          origin: origin(),
+          id: id(),
+          instance: instance()
+        }
+
+  defstruct [:height, :weight, :id, :instance, :origin, hashes: %{}]
+
+  @impl true
+  def new(args \\ [])
+  def new(attrs) when is_map(attrs), do: struct(%__MODULE__{}, attrs)
+  def new(attrs) when is_list(attrs), do: struct(%__MODULE__{}, attrs)
+  def new(_attrs), do: {:error, [args: :invalid_format]}
+
+  @spec set_hashes(cut :: t(), hashes :: hashes()) :: result()
+  def set_hashes(%__MODULE__{} = cut, %{} = hashes) do
+    keys = Map.keys(hashes)
+    validate_hashes(cut, keys, hashes)
+  end
+
+  def set_hashes(%__MODULE__{}, _hashes), do: {:error, [hashes: :not_a_map]}
+
+  @spec add_hash(cut :: t(), chain_id :: chain_id(), hash :: map()) :: result()
+  def add_hash(
+        %__MODULE__{hashes: hashes} = cut,
+        chain_id,
+        %{hash: hash_value, height: height} = hash
+      ) do
+    with {:ok, chain_id} <- validate_chain_id(chain_id),
+         true <- is_binary(hash_value),
+         true <- height >= 0 do
+      %{cut | hashes: Map.put(hashes, chain_id, hash)}
+    else
+      _ -> {:error, [args: :invalid]}
+    end
+  end
+
+  def add_hash(%__MODULE__{}, _chain_id, _hash), do: {:error, [args: :invalid]}
+
+  @spec remove_hash(cut :: t(), chain_id :: chain_id()) :: result()
+  def remove_hash(%__MODULE__{hashes: hashes} = cut, chain_id) when chain_id in 0..19,
+    do: %{cut | hashes: Map.delete(hashes, String.to_atom("#{chain_id}"))}
+
+  def remove_hash(_cut, _chain_id), do: {:error, [chain_id: :invalid]}
+
+  @spec set_height(cut :: t(), height :: height()) :: result()
+  def set_height(%__MODULE__{} = cut, height) when is_integer(height), do: %{cut | height: height}
+  def set_height(%__MODULE__{}, _height), do: {:error, [height: :not_an_integer]}
+
+  @spec set_weight(cut :: t(), weight :: weight()) :: result()
+  def set_weight(%__MODULE__{} = cut, weight) when is_binary(weight), do: %{cut | weight: weight}
+  def set_weight(%__MODULE__{}, _weight), do: {:error, [weight: :not_a_string]}
+
+  @spec set_origin(cut :: t(), origin :: origin()) :: result()
+  def set_origin(%__MODULE__{} = cut, %{id: id, address: address} = origin) when is_map(origin) do
```

Comment:
Please remove the guard, it is not necessary, as pattern matching already does this